keps/993-two-phase-admission/README.md

```diff
@@ -253,6 +254,27 @@ The controller implementing a particular check should:
 * After approving the workload, keep an eye on the check if it starts failing,
   fail the check and cause the workload to move back to the suspended state.
 
+### Preemption Admission Check Controller
```

Comment:
I have an alternative solution (please consider it, even though we might still go with your proposed solution):

Once a workload is allowed to preempt, we put it back in the scheduling queue. When poped from the queue, this workload uses a different snapshot: one that doesn't have the workload with quota reserved but not admitted.
Then we run the same algorithm we have, including preemption.

- If the checks all passed, then we can admit the workload.
- Otherwise, it goes back into the queue, probably into the "inadmissible" staging area and it can get poped back once all checks have passed.

This has the advantage that a single component is "admitting" workloads and issuing preemptions, reducing the risk for race conditions. In your proposed approach, it is possible that the preemption controller determines that a workload is admissible, at the same time that the scheduler is admitting another workload that doesn't require preemptions or checks. And the workloads that require preemption are not prioritized by the same rules as the ones that go through the scheduling queue.

Reply:
> Once a workload is allowed to preempt, we put it back in the scheduling queue. 
The problem is that there is no guarantee that the same assignment will be available.

> This has the advantage that a single component is "admitting" workloads and issuing preemption, reducing the risk for race conditions. 
The problem is that the triggers the scheduler should respond to and the type of tasks it needs to do will dramatically increase. (I did think of having the scheduler more involved when I first started working on two-step admission but keeping the scheduler involved in only managing quota reservation is the best option in my opinion).

>  In your proposed approach, it is possible that the preemption controller determines that a workload is admissible, at the same time that the scheduler is admitting another workload that doesn't require preemptions or checks.
Actually the quota reserved by all the preemption pending  workloads is "in use" from the schedulers point of view.

>  And the workloads that require preemption are not prioritized by the same rules as the ones that go through the scheduling queue.
Indeed when it comes to the time when they are issuing eviction ,  and this one of the purpose of having preemption policies. However to get in the preemption pending state the rules are the same and during the preemption pending they are still  eligible to be evicted by a higher prio workload.

Reply:
> The problem is that there is no guarantee that the same assignment will be available.

The same problem would be faced by a preemption controller. It needs to check that the assignment is still viable. We can have a mechanism in the scheduler logic to first try the assignment that was already calculated.

> Actually the quota reserved by all the preemption pending workloads is "in use" from the schedulers point of view

True.

> However to get in the preemption pending state the rules are the same and during the preemption pending they are still eligible to be evicted by a higher prio workload.

If two workloads have their admission checks completed, there is no guarantee that both will be able to fit after issuing preemptions. Therefore, we should prioritize them. Sure, the higher priority one could still preempt the lower priority one, but the disruption could be avoided if we had prioritized preemptions.

It's almost starting to sound like we should have a a second "preemption scheduler", instead of a "preemption controller"

Reply:
> "preemption scheduler"

@alculquicondor Does this mean we should introduce a scheduler for preemption separate from an existing scheduler (admission scheduler?)?

I'm concerned that two schedulers can work in asynchronous cooperation as we expected.

I was thinking about the separate scheduler when I reviewed @trasc 's implementation PR. 
However, in the PR, I said we might want to somehow implement the preemption logic to the scheduling cycle instead of introducing a new scheduler.

Because I thought that it might be hard to handle different schedulers well due to some conflicts, such as lower being preempted by higher, as you mentioned. Moreover, I'm wondering if the issue would be more complex when the borrowing semantics would be related to the scheduling logic.

Honestly, I don't have a clear answer. However, I'm wondering if we should discuss the second scheduler for the preemption focused on the existing preemption logic, not including two-phase admission once away from two-phase admission with preemption.

Reply:
I'm already thinking of leaving this change out of release-0.5, as it's highly complex and we don't have clear answers yet.

That said, I think the idea of two schedulers isn't that bad, based on:

> Actually the quota reserved by all the preemption pending workloads is "in use" from the schedulers point of view

So they might not conflict with each other. Still, it might be valuable to revisit the idea that a single scheduler has the two modes: preempting and non-preempting.

Let's give it a thought to all these ideas and come back to it in a couple of weeks.

Reply:
> That said, I think the idea of two schedulers isn't that bad, based on:

I don't have a strong objection regarding two schedulers. I just wanted to say that I'm wondering if we shouldn't move this forward in a hurry.

> I'm already thinking of leaving this change out of release-0.5, as it's highly complex and we don't have clear answers yet.

I have the completely same opinion.

Reply:
This is just an idea, as a byproduct, If we introduce a preemption scheduler separate from an existing scheduler, it might make providing [pluggable queue policies](https://github.com/kubernetes-sigs/kueue/issues/10) much easier since we might separate preemption logic from the flavorassigner.

Reply:
> If two workloads have their admission checks completed, there is no guarantee that both will be able to fit after issuing preemptions. Therefore, we should prioritize them.

The priority is taken into account when creating the updated list of eviction candidates. 

Say you have a queue with capacity 5cpu and :
wl0 (prio 0,  3cpu, admitted )
wl1 (prio 1, 3cpu, quota reserved, ready for preemption)
wl2 (prio 2, 3cpu, quota reserved, ready for preemption)
During the preemption check the preemption check controller will try to:

- Get the list of candidates for wl1 while the queue has wl0 and wl2 using quota, since wl2 has higher priority the list will be empty, therefore the preemption failed for wl1. (check state `false`)
- Get the list of candidates for wl2 while the queue has wl0 and wl1 using quota, the list will contain (wl1 and wl2) they will be marked for eviction.
 
(the order in which the steps above are done is irrelevant)

I a future run (after the eviction are done) wl2 will fit. (check state `true`)

> I'm already thinking of leaving this change out of release-0.5, as it's highly complex and we don't have clear answers yet.

We have a working implemetation.

Reply:
> (the order in which the steps above are done is irrelevant)

After reviewing your argument, I agree with this assessment.

> We have a working implemetation.

The problem is that we are touching currently stable functionality in ways that are hard to predict the outcomes. We might need to evaluate where our coverage is weak, so we can add it before changing the logic. We can't merge such a big PR lightly.

But now we are at a good point in the release cycle in which we can proceed with such an overtake. I still need a little more time to review.